app/controllers/CampaignController.scala

```diff
@@ -11,25 +14,40 @@ import wiring.AppComponents.Stage
 
 import scala.concurrent.{ExecutionContext, Future}
 
-class CampaignController(googleAuthAction: GoogleAuthenticatedAction, service: JsonDynamoService[Campaign, Future], stage: Stage, implicit val ec: ExecutionContext) {
+class CampaignController(googleAuthAction: GoogleAuthenticatedAction, campaignService: JsonDynamoService[Campaign, Future], promotionService: JsonDynamoService[AnyPromotion, Future], stage: Stage, implicit val ec: ExecutionContext) {
 
   def all(group: Option[String]) = googleAuthAction.async {
-    service.all.map(campaigns => {
-      val filtered = group.flatMap(CampaignGroup.fromId).fold(campaigns)(group => campaigns.filter(_.group == group))
-      Ok(Json.toJson(filtered.sortBy(_.name)))
-    })
+    import com.gu.memsub.promo.Formatters.PromotionFormatters._
+
+    val now = DateTime.now
+    val campaignsF = campaignService.all
+    val promotionsF = promotionService.all
+    for {
+      campaigns <- campaignsF
+      promotions <- promotionsF
+    } yield {
+      val filtered = group.flatMap(CampaignGroup.fromId).map(group => campaigns.filter(_.group == group)) getOrElse campaigns
```

Comment:
From a style point of view, I'm not very keen on these large yield blocks. 

I think this code would be much easier to read (and unit test) if we split this out into some smaller functions.